### PR TITLE
ApiDoc for Lock, RingBuffer and Set

### PR DESCRIPTION
**Pull request by oguzdemir**
ApiDoc added to lock.py , ringbuffer.py and set.py with Python Sphinx format.

add_entry_listener method of set class is refactored for indicating parameters which are functions.

### CODE REVIEW

File: hazelcast/proxy/lock.py [hunk 1]
```diff
@@ -6,29 +6,82 @@
...
+
+        :return: (Bool), true if this lock is locked by current thread, false otherwise.
+        """
         return self._encode_invoke(lock_is_locked_by_current_thread_codec,
                                    thread_id=thread_id())
 
     def lock(self, lease_time=-1):
+        """
+        Acquires the lock, If a lease time is specified, lock will be released after this lease time.
+
```
Comment (Serdaro) [inline hazelcast/proxy/lock.py]: Acquires the lock. If a lease time is....

File: hazelcast/proxy/lock.py [hunk 2]
```diff
@@ -6,29 +6,82 @@
...
+        If the lock is not available then the current thread becomes disabled for thread scheduling purposes and lies
+        dormant until the lock has been acquired.
+
```
Comment (Serdaro) [inline hazelcast/proxy/lock.py]: If the lock is not available, the current thread....

File: hazelcast/proxy/lock.py [hunk 3]
```diff
@@ -6,29 +6,82 @@
...
+            dormant until one of the followings happens:
+                * the lock is acquired by the current thread, or
+                * the specified waiting time elapses.
+
+        If lease time is provided, lock will be released after this time elapses.
+
+        :param timeout: (Long), maximum time in seconds to wait for the lock (optional).
+        :param lease_time: (Long), time in seconds to wait before releasing the lock (optional).
+        :return: (Bool), true if the lock was acquired and otherwise, false
+        """
```
Comment (Serdaro) [inline hazelcast/proxy/lock.py]: otherwise, false.

File: hazelcast/proxy/ringbuffer.py [hunk 4]
```diff
@@ -11,9 +11,30 @@
 
 
 class Ringbuffer(PartitionSpecificProxy):
+    """
+    A Ringbuffer is a data-structure where the content is stored in a ring like structure. A ringbuffer has a capacity
+    so it won't grow beyond that capacity and endanger the stability of the system. If that capacity is exceeded, than
+    the oldest item in the ringbuffer is overwritten. The ringbuffer has 2 always incrementing sequences:
+        #. tail_sequence: this is the side where the youngest item is found. So the tail is the side of the ringbuffer
```
Comment (Serdaro) [inline hazelcast/proxy/ringbuffer.py]: (To be consistent, let's use "Ringbuffer" for the rest. (not ring buffer, ringbuffer or RingBuffer).

File: hazelcast/proxy/ringbuffer.py [hunk 5]
```diff
@@ -23,21 +44,74 @@ def cache_capacity(f):
...
 
     def add(self, item, overflow_policy=OVERFLOW_POLICY_OVERWRITE):
+        """
+        Adds the specified item to the tail of the ringbuffer. If there is no space in the ringbuffer, the action is determined by
+        overflow policy as follows:
+            #. OVERFLOW_POLICY_OVERWRITE (default): oldest item in the ringbuffer is overwritten and item is added.
+            #. OVERFLOW_POLICY_FAIL: call returns -1 and add fails.
+        The reason that FAIL exist is to give the opportunity to obey the ttl. If blocking behavior is required, this
+        can be implemented using retrying in combination with a exponential backoff.
+            >>> sleepMS = 100;
```
Comment (Serdaro) [inline hazelcast/proxy/ringbuffer.py]: ....with an exponential...

File: hazelcast/proxy/ringbuffer.py [hunk 6]
```diff
@@ -23,21 +44,74 @@ def cache_capacity(f):
...
+        :param item: (Object), the specified item to be added.
+        :param overflow_policy: (int), the OverflowPolicy to be used when there is no space (optional).
+        :return: (Long), the sequenceId of the added item, or -1 if the add failed.
+        """
         return self._encode_invoke(ringbuffer_add_codec, value=self._to_data(item), overflow_policy=overflow_policy)
 
     def add_all(self, items, overflow_policy=OVERFLOW_POLICY_OVERWRITE):
+        """
+        Adds all of the item in specified collection to the tail of the ringbuffer. A add_all is likely to outperform
+        multiple calls to add(Object) due to better io utilization and a reduced number of executed operations. The
```
Comment (Serdaro) [inline hazelcast/proxy/ringbuffer.py]: Adds all of the items in the specified.......An add_all is.....

File: hazelcast/proxy/set.py [hunk 7]
```diff
@@ -40,26 +63,41 @@ def handle_event_item(item, uuid, event_type):
...
 
         return self._start_listening(request,
                                      lambda m: set_add_listener_codec.handle(m, handle_event_item),
                                      lambda r: set_add_listener_codec.decode_response(r)['response'],
                                      self.partition_key)
 
     def clear(self):
+        """
+        Clears the set, set will be empty with this call.
+        """
```
Comment (Serdaro) [inline hazelcast/proxy/set.py]: Clears the set. Set will be.....

File: hazelcast/proxy/set.py [hunk 8]
```diff
@@ -40,26 +63,41 @@ def handle_event_item(item, uuid, event_type):
...
+        :return: (Bool), true if the specified item exists in this set, false otherwise.
+        """
         check_not_none(item, "Value can't be None")
         item_data = self._to_data(item)
         return self._encode_invoke(set_contains_codec, value=item_data)
 
     def contains_all(self, items):
+        """
+        Determines whether this set contains all of the items in specified collection or not.
+
```
Comment (Serdaro) [inline hazelcast/proxy/set.py]: ....items in the specified....

File: hazelcast/proxy/set.py [hunk 9]
```diff
@@ -40,26 +63,41 @@ def handle_event_item(item, uuid, event_type):
...
+        :param items: (Collection), the specified collection which includes the items to be searched.
+        :return: (Bool), true if all of the items in specified collection exist in this set, false otherwise.
+        """
```
Comment (Serdaro) [inline hazelcast/proxy/set.py]: ...in the specified collection...

File: hazelcast/proxy/set.py [hunk 10]
```diff
@@ -87,9 +147,22 @@ def remove_all(self, items):
...
+        :param registration_id: (String), id of the listener to be deleted.
+        :return: (Bool), true if the item listener is removed, false otherwise.
+        """
         return self._stop_listening(registration_id, lambda i: set_remove_listener_codec.encode_request(self.name, i))
 
     def retain_all(self, items):
+        """
+        Removes the items which are not contained in the specified collection. In other words, that only the items that
+        are contained in the specified collection will be retained.
+
```
Comment (Serdaro) [inline hazelcast/proxy/set.py]: In other words, only the items that.....

**Comment by asimarslan** [comment]
run-lab-run

**Pull request merged**